wolfcrypt/src/chacha20_poly1305.c

```diff
@@ -248,4 +253,167 @@ static void word32ToLittle64(const word32 inLittle32, byte outLittle64[8])
+    }
+
+    if (ret == 0) {
+        aead->state = CHACHA20_POLY1305_STATE_READY;
+    }
+
+    return ret;
+}
+
+/* optional additional authentication data */
+int wc_ChaCha20Poly1305_UpdateAad(ChaChaPoly_Aead* aead,
+    const byte* inAAD, word32 inAADLen)
+{
+    int ret = 0;
+
+    if (aead == NULL || inAAD == NULL) {
+        return BAD_FUNC_ARG;
+    }
+    if (aead->state != CHACHA20_POLY1305_STATE_READY &&
+        aead->state != CHACHA20_POLY1305_STATE_AAD) {
+        return BAD_STATE_E;
+    }
+
+    if (inAADLen > 0) {
+        ret = wc_Poly1305Update(&aead->poly, inAAD, inAADLen);
+        if (ret == 0) {
+            aead->aadLen += inAADLen;
+            aead->state = CHACHA20_POLY1305_STATE_AAD;
+        }
+    }
+
+    return ret;
+}
+
+int wc_ChaCha20Poly1305_UpdateData(ChaChaPoly_Aead* aead,
+    byte* data, word32 dataLen)
+{
+    int ret = 0;
+
+    if (aead == NULL || data == NULL) {
+        return BAD_FUNC_ARG;
+    }
+    if (aead->state != CHACHA20_POLY1305_STATE_READY &&
+        aead->state != CHACHA20_POLY1305_STATE_AAD &&
+        aead->state != CHACHA20_POLY1305_STATE_DATA) {
+        return BAD_STATE_E;
+    }
+
+    if (aead->state == CHACHA20_POLY1305_STATE_AAD) {
+        /* Pad the AAD to 16 bytes */
+        byte padding[CHACHA20_POLY1305_MAC_PADDING_ALIGNMENT - 1];
+        word32 paddingLen = -(int)aead->aadLen &
+            (CHACHA20_POLY1305_MAC_PADDING_ALIGNMENT - 1);
+        if (paddingLen > 0) {
+            XMEMSET(padding, 0, paddingLen);
+            ret = wc_Poly1305Update(&aead->poly, padding, paddingLen);
+        }
+    }
+    aead->state = CHACHA20_POLY1305_STATE_DATA;
+
+    if (ret == 0) {
+        /* Perform ChaCha20 encrypt or decrypt inline and Poly1305 auth calc */
+        if (aead->isEncrypt) {
+            ret = wc_Chacha_Process(&aead->chacha, data, data, dataLen);
+            if (ret == 0)
+                ret = wc_Poly1305Update(&aead->poly, data, dataLen);
+        }
+        else {
+            ret = wc_Poly1305Update(&aead->poly, data, dataLen);
+            if (ret == 0)
+                ret = wc_Chacha_Process(&aead->chacha, data, data, dataLen);
+        }
+    }
+    if (ret == 0) {
+        aead->dataLen += dataLen;
+    }
+    return ret;
+}
+
+int wc_ChaCha20Poly1305_Final(ChaChaPoly_Aead* aead,
+    byte outAuthTag[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE])
+{
+    int ret = 0;
+    byte padding[CHACHA20_POLY1305_MAC_PADDING_ALIGNMENT - 1];
+    word32 paddingLen;
+    byte little64[16]; /* word64 * 2 */
+
+    if (aead == NULL || outAuthTag == NULL) {
+        return BAD_FUNC_ARG;
+    }
+    if (aead->state != CHACHA20_POLY1305_STATE_DATA) {
+        return BAD_STATE_E;
+    }
+    aead->state = CHACHA20_POLY1305_STATE_FINAL;
+
+    /* Pad the ciphertext to 16 bytes */
+    paddingLen = -(int)aead->dataLen &
+        (CHACHA20_POLY1305_MAC_PADDING_ALIGNMENT - 1);
+    if (paddingLen > 0) {
+        XMEMSET(padding, 0, paddingLen);
```

Comment:
Move 'XMEMSET(padding, 0, paddingLen);' to the beginning of the function.

Reply:
I cannot move this because paddingLen will not be set. I could move to top and replace with `XMEMSET(padding, 0, sizeof(padding));`. Is this a DO-178 requirement?

Reply:
You are correct. Use XMEMSET(padding, 0, sizeof(padding)); 
consistency and accuracy is important for DO-178.

Reply:
I moved things around in commit https://github.com/wolfSSL/wolfssl/commit/bff6dcec248cb3e7a67d037a1b7b6db606981cbf and this should be improved. Let me know if its not DO-178 okay...

Reply:
@dgarske, it helps. a variable declaration such as "byte padding[]" that is in the middle of the function might end up being cert defects that we may need to handle later.